File: backend/models/media_edge.go
package models

import pb "goodanime-backend/proto"

// ToProto converts MediaEdge to its protobuf representation.
func (e *MediaEdge) ToProto() *pb.MediaEdge {
	if e == nil {
		return nil
	}
	return &pb.MediaEdge{
		RelationType: e.RelationType,
		Node:         e.Node.ToProto(),
	}
}

// ToProto converts MediaConnection to its protobuf representation.
func (c *MediaConnection) ToProto() *pb.MediaConnection {
	if c == nil {
		return nil
	}
	res := &pb.MediaConnection{}
	for _, edge := range c.Edges {
		res.Edges = append(res.Edges, edge.ToProto())
	}
	return res
}

// MediaEdge represents a link between two media
type MediaEdge struct {
	RelationType string    `json:"relationType"`
	Node         *MediaMin `json:"node"`
}

// MediaConnection represents a paginated list of media relations
type MediaConnection struct {
	Edges []MediaEdge `json:"edges"`
}

File: lib/models/media_edge.dart
import '../proto/media_edge.pb.dart' as pb;
import '../proto/media_min.pb.dart' as pbm;
import 'media_min.dart';

/// Represents a link between two media
class MediaEdge {
  /// Relation type
  final String relationType;

  /// The related media
  final MediaMin? node;

  /// Creates a media edge
  const MediaEdge({required this.relationType, this.node});

  /// Creates a media edge from a protobuf object
  factory MediaEdge.fromProto(pb.MediaEdge pbObj) {
    return MediaEdge(
      relationType: pbObj.relationType,
      node: pbObj.hasNode() ? MediaMin.fromProto(pbObj.node) : null,
    );
  }

  /// Converts the media edge to a protobuf object
  pb.MediaEdge toProto() {
    return pb.MediaEdge(
      relationType: relationType,
      node: node?.toProto() as pbm.MediaMin,
    );
  }
}

/// Represents a paginated list of media relations
class MediaConnection {
  /// Relation edges
  final List<MediaEdge> edges;

  /// Creates a media connection
  const MediaConnection({required this.edges});

  /// Creates a media connection from a protobuf object
  factory MediaConnection.fromProto(pb.MediaConnection pbObj) {
    return MediaConnection(
      edges: pbObj.edges.map(MediaEdge.fromProto).toList(),
    );
  }

  /// Converts the media connection to a protobuf object
  pb.MediaConnection toProto() {
    return pb.MediaConnection(edges: edges.map((e) => e.toProto()).toList());
  }
}
